### DISCUSSION TITLE
npm or yarn cache to speed up the deployment

### DISCUSSION THREAD

**Discussion by mohdmusthafa** [coollabsio/coolify#3309] [Q&A]
I have a project as monorepo (turbo repo) + yarn. I have 8 packages in the repo and I filter out the installations using `yarn workspace api install`, `yarn workspace admin install` etc. the issue is that yarn install is taking too much time for each deployment as it needs to download +700mb of packages, is there any way that I can cache the dependencies to reuse in the deployment

Note:
1. docker cleanup enabled with frequency (*/10 * * * *)
2. packages tend to use same packages
3. I am looking for something like we have on the github actions (actions/cache@v4)
4. If attaching a volume to yarn cache directory helps, please elaborate on how to achieve that (I'm new to this..:)

**Comment by namdien177** [comment]
I was trying to use the persistence storage feature to achieve what you said as well, but that feature seems to mount the directory **AFTER** the docker image finishes the run, so nothing much can be done there.

If we can have persistence storage with a "checkbox" to mount before the build, then it's perfect, IMO.

(my Raspberry Pi 4 takes like ~1500 seconds just for `bun install`. I would love to have `.bun` and `node_modules` cached somehow.)

**Comment by darklight9811** [comment]
Would love to have this too, bun install is taking around 2 minutes

**Comment by hades200082** [comment]
With nextjs + payload + mastra deployments take nearly 15 minutes. This is an essential feature request.

**Reply by djsisson** [reply]
run 
```shell
docker buildx create --name test --driver docker-container --use
```
this will create a build container that uses a volume for caching, so it will not get effected by coolify docker cleanup, please note this volume defaults to reserving 30% of diskspace, you can configure it in options or with config

https://docs.docker.com/reference/cli/docker/buildx/create/#buildkitd-config

So now your build use this buildx container, you just need to mount your cache in your docker files

a rough example

```dockerfile
# syntax=docker/dockerfile:1

#### Build Stage ####
FROM node:23-alpine AS builder

# Install necessary build dependencies
RUN apk add --no-cache libc6-compat

# Set working directory
WORKDIR /app

# Copy package manifests
COPY package.json package-lock.json ./

# Node Env (prodution to skip dev dependencies)
ENV NODE_ENV=production

# Install dependencies with caching
RUN --mount=type=cache,target=/root/.npm \
    npm ci

# Copy remaining source code
COPY . .

# Build Args for static generation
ARG NEXT_PUBLIC_XXXX

# Build in standalone mode
RUN npm run build

#### Production Stage ####
FROM node:23-alpine AS runner

# Node Env
ENV NODE_ENV=production

# Create non-root user and group
RUN addgroup -g 1001 node \
 && adduser -S -u 1001 -G node nextjs

# Set working directory
WORKDIR /app

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000
ENV HOSTNAME="0.0.0.0"
CMD ["node", "server.js"]
```

you can cache apt(bookworm etc) turbo and packages using cache mount, you can specify ```id``` if you build multiple projects and don't want them to share cache, as it defaults to ```id = path```
ps. this assumes you're running npm and standalone in nextjs

each framework has different caching strategies, so you would need to change cache mounts as required

e.g for go its

```dockerfile
RUN --mount=type=cache,target=/go/pkg/mod/ \
    --mount=type=bind,source=go.mod,target=go.mod \
    --mount=type=bind,source=go.sum,target=go.sum \
    go mod download -x

ARG TARGETARCH
ARG TARGETOS

RUN --mount=type=cache,target=/go/pkg/mod/ \
    --mount=type=bind,target=. \
    --mount=type=cache,target=/root/.cache/go-build \
    CGO_ENABLED=0 GOARCH=$TARGETARCH GOOS=$TARGETOS go build -ldflags="-s -w" -o /usr/bin/server ./cmd/main
```

**Reply by theDanielJLewis** [reply]
That seems interesting, but how would we do this when we're running applications _without_ custom docker or compose files?

**Reply by ntgussoni** [reply]
eg: on nixpacks

**Reply by djsisson** [reply]
nixpacks already uses the same mount caches in the generated Dockerfile
but it uses an id so that they are scoped only to that id (to keep caches isolated to each repo), which is some kind of hash which includes the build path

but i believe each deployment is given its own deployment uuid by coolify, therefore every build path is different, so the id will be different each time so there will be no cache to reuse and the build cache just gets bigger overtime

https://github.com/coollabsio/coolify/blob/96b3e269c57f332b48a6bd00f97c25492e2ec13a/app/Jobs/ApplicationDeploymentJob.php#L214-L215

https://github.com/coollabsio/coolify/blob/96b3e269c57f332b48a6bd00f97c25492e2ec13a/app/Models/Application.php#L970-L973
where uuid is the deployment uuid

when running commands inside the coolify helper there is no need to change the base dir everytime since the container is isolated so this is something that should absolutely be changed

**Reply by ntgussoni** [reply]
@andrasbacsai is there any workaround we can do or way to help make a change ?

**Reply by gsporto** [reply]
Having a functional cache for pnpm, turborepo, tsc would go long ways for build time, also will probably help with build cache in turpopack and vite

**Reply by andrasbacsai** [reply]
> nixpacks already uses the same mount caches in the generated Dockerfile but it uses an id so that they are scoped only to that id (to keep caches isolated to each repo), which is some kind of hash which includes the build path
> 
> but i believe each deployment is given its own deployment uuid by coolify, therefore every build path is different, so the id will be different each time so there will be no cache to reuse and the build cache just gets bigger overtime
> 
> https://github.com/coollabsio/coolify/blob/96b3e269c57f332b48a6bd00f97c25492e2ec13a/app/Jobs/ApplicationDeploymentJob.php#L214-L215
> 
> https://github.com/coollabsio/coolify/blob/96b3e269c57f332b48a6bd00f97c25492e2ec13a/app/Models/Application.php#L970-L973
> 
> 
> where uuid is the deployment uuid
> when running commands inside the coolify helper there is no need to change the base dir everytime since the container is isolated so this is something that should absolutely be changed

tldr: Coolify have cache in place for most of the builds. **If you have the automated cleanup set it will cleanup build caches as well.** 


For example, Nixpacks has two caches. 

1. Docker layer cache (caches the nixarchives)
2. Nixpacks cache (`--cache-key`). Coolify sets the `--cache-key` to be the application id, so it does not matter if the deployment uuid changes. It caches npm/pnpm/yarn artifacts.

Example build where I only changed the code of the applications (not package.json or nixarchives dependencies):
```
[CMD]: docker exec dsksg4kk8sk04s8wo0ocw08o bash -c 'nixpacks build -c /artifacts/thegameplan.json --cache-key 'akwk0oscw4woww88gcs04o4w' --no-error-without-start -n akwk0oscw4woww88gcs04o4w:0de85455a63c48d63aa9aeec367065890f2ced4e-build /artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa -o /artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa'
╔══════════════ Nixpacks v1.39.0 ═════════════╗
║ setup      │ nodejs_18, npm-9_x, curl, wget ║
║─────────────────────────────────────────────║
║ install    │ npm ci                         ║
║─────────────────────────────────────────────║
║ build      │ npm run build                  ║
║─────────────────────────────────────────────║
║ start      │ npm run start                  ║
╚═════════════════════════════════════════════╝
Saved output to:
/artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa
[CMD]: docker exec dsksg4kk8sk04s8wo0ocw08o bash -c 'cat /artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa/.nixpacks/Dockerfile'
FROM ghcr.io/railwayapp/nixpacks:ubuntu-1745885067
ENTRYPOINT ["/bin/bash", "-l", "-c"]
WORKDIR /app/
COPY .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix
RUN nix-env -if .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix && nix-collect-garbage -d
RUN sudo apt-get update && sudo apt-get install -y --no-install-recommends curl wget
ARG CI COOLIFY_BRANCH COOLIFY_CONTAINER_NAME COOLIFY_FQDN COOLIFY_RESOURCE_UUID COOLIFY_URL NEXT_PUBLIC_ANALYTICS_ID NIXPACKS_METADATA NODE_ENV NPM_CONFIG_PRODUCTION SOURCE_COMMIT asd
ENV CI=$CI COOLIFY_BRANCH=$COOLIFY_BRANCH COOLIFY_CONTAINER_NAME=$COOLIFY_CONTAINER_NAME COOLIFY_FQDN=$COOLIFY_FQDN COOLIFY_RESOURCE_UUID=$COOLIFY_RESOURCE_UUID COOLIFY_URL=$COOLIFY_URL NEXT_PUBLIC_ANALYTICS_ID=$NEXT_PUBLIC_ANALYTICS_ID NIXPACKS_METADATA=$NIXPACKS_METADATA NODE_ENV=$NODE_ENV NPM_CONFIG_PRODUCTION=$NPM_CONFIG_PRODUCTION SOURCE_COMMIT=$SOURCE_COMMIT asd=$asd
# setup phase
# noop
# install phase
ENV NIXPACKS_PATH=/app/node_modules/.bin:$NIXPACKS_PATH
COPY . /app/.
RUN --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-/root/npm,target=/root/.npm npm ci
# build phase
COPY . /app/.
RUN --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-next/cache,target=/app/.next/cache --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-node_modules/cache,target=/app/n-ode_modules/.cache npm run build
RUN printf '\nPATH=/app/node_modules/.bin:$PATH' >> /root/.profile
# start
COPY . /app
CMD ["npm run start"]
[CMD]: docker exec dsksg4kk8sk04s8wo0ocw08o bash -c 'cat /artifacts/build.sh'
docker build --add-host akwk0oscw4woww88gcs04o4w-pr-49:192.168.97.13 --add-host coolify:192.168.97.11 --add-host coolify-db:192.168.97.7 --add-host coolify-mail:192.168.97.6 --add-host coolify-minio:192.168.97.3 --add-host coolify-realtime:192.168.97.4 --add-host coolify-redis:192.168.97.5 --add-host coolify-testing-host:192.168.97.10 --add-host coolify-vite-1:192.168.97.9 --add-host ioogkco4g8cwk40gc0ooosg8-pr-41:192.168.97.12 --add-host n8wssw8wggw4g84soc8co8wk-pr-49:192.168.97.19 --add-host ugcc04o0sc4o4so8cgcw8sg0:192.168.97.2 --network host -f /artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa/.nixpacks/Dockerfile --progress plain -t akwk0oscw4woww88gcs04o4w:0de85455a63c48d63aa9aeec367065890f2ced4e-build --build-arg CI='true' --build-arg COOLIFY_BRANCH='v4.x' --build-arg COOLIFY_CONTAINER_NAME='akwk0oscw4woww88gcs04o4w-081435880083' --build-arg COOLIFY_FQDN='akwk0oscw4woww88gcs04o4w.127.0.0.1.sslip.io' --build-arg COOLIFY_RESOURCE_UUID='akwk0oscw4woww88gcs04o4w' --build-arg COOLIFY_URL='http://akwk0oscw4woww88gcs04o4w.127.0.0.1.sslip.io'/ --build-arg NIXPACKS_METADATA='node' --build-arg NODE_ENV='production' --build-arg NPM_CONFIG_PRODUCTION='false' --build-arg SOURCE_COMMIT='0de85455a63c48d63aa9aeec367065890f2ced4e' --build-arg NEXT_PUBLIC_ANALYTICS_ID='prod' --build-arg asd='asd' --build-arg 'COOLIFY_URL=http://akwk0oscw4woww88gcs04o4w.127.0.0.1.sslip.io'/ --build-arg 'COOLIFY_FQDN=akwk0oscw4woww88gcs04o4w.127.0.0.1.sslip.io' --build-arg 'COOLIFY_BRANCH=v4.x' --build-arg 'COOLIFY_RESOURCE_UUID=akwk0oscw4woww88gcs04o4w' --build-arg 'COOLIFY_CONTAINER_NAME=akwk0oscw4woww88gcs04o4w-081435880083' /artifacts/dsksg4kk8sk04s8wo0ocw08o/nextjs/spa
[CMD]: docker exec dsksg4kk8sk04s8wo0ocw08o bash -c 'bash /artifacts/build.sh'
#0 building with "default" instance using docker driver
#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 1.56kB done
#1 DONE 0.1s
#2 [internal] load metadata for ghcr.io/railwayapp/nixpacks:ubuntu-1745885067
#2 DONE 0.5s
#3 [internal] load .dockerignore
#3 transferring context: 2B done
#3 DONE 0.0s
#4 [stage-0  1/11] FROM ghcr.io/railwayapp/nixpacks:ubuntu-1745885067@sha256:d45c89d80e13d7ad0fd555b5130f22a866d9dd10e861f589932303ef2314c7de
#4 DONE 0.0s
#5 [internal] load build context
#5 transferring context: 250.88kB done
#5 DONE 0.0s
#6 [stage-0  2/11] WORKDIR /app/
#6 CACHED
#7 [stage-0  3/11] COPY .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix
#7 CACHED
#8 [stage-0  4/11] RUN nix-env -if .nixpacks/nixpkgs-ffeebf0acf3ae8b29f8c7049cd911b9636efd7e7.nix && nix-collect-garbage -d
#8 CACHED
#9 [stage-0  5/11] RUN sudo apt-get update && sudo apt-get install -y --no-install-recommends curl wget
#9 CACHED
#10 [stage-0  6/11] COPY . /app/.
#10 DONE 0.0s
#11 [stage-0  7/11] RUN --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-/root/npm,target=/root/.npm npm ci
#11 0.100 npm warn config production Use `--omit=dev` instead.
#11 4.756
#11 4.756 added 347 packages, and audited 348 packages in 5s
#11 4.756
#11 4.756 140 packages are looking for funding
#11 4.756   run `npm fund` for details
#11 4.780
#11 4.780 2 vulnerabilities (1 low, 1 moderate)
#11 4.780
#11 4.780 To address all issues, run:
#11 4.780   npm audit fix
#11 4.780
#11 4.780 Run `npm audit` for details.
#11 DONE 5.7s
#12 [stage-0  8/11] COPY . /app/.
#12 DONE 0.0s
#13 [stage-0  9/11] RUN --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-next/cache,target=/app/.next/cache --mount=type=cache,id=akwk0oscw4woww88gcs04o4w-node_modules/cache,target=/app/node_modules/.cache npm run build
#13 0.072 npm warn config production Use `--omit=dev` instead.
#13 0.082
#13 0.082 > nextjs-spa@0.0.0 build
#13 0.082 > next build
#13 0.082
#13 0.361    ▲ Next.js 15.3.4
#13 0.362
#13 0.379    Creating an optimized production build ...
#13 2.470  ✓ Compiled successfully in 0ms
#13 2.472    Linting and checking validity of types ...
#13 3.342    Collecting page data ...
#13 3.802    Generating static pages (0/5) ...
#13 4.092    Generating static pages (1/5)
#13 4.092    Generating static pages (2/5)
#13 4.092    Generating static pages (3/5)
#13 4.092  ✓ Generating static pages (5/5)
#13 4.254    Finalizing page optimization ...
#13 4.254    Collecting build traces ...
#13 6.858    Exporting (0/3) ...
#13 7.080  ✓ Exporting (3/3)
#13 7.087
#13 7.090 Route (app)                                 Size  First Load JS
#13 7.090 ┌ ○ /                                      135 B         101 kB
#13 7.090 └ ○ /_not-found                            978 B         102 kB
#13 7.090 + First Load JS shared by all             101 kB
#13 7.090   ├ chunks/4bd1b696-67ee12fb04071d3b.js  53.2 kB
#13 7.090   ├ chunks/684-2d117c50cbadadca.js         46 kB
#13 7.090   └ other shared chunks (total)          1.88 kB
#13 7.090
#13 7.091
#13 7.091 ○  (Static)  prerendered as static content
#13 7.091
#13 DONE 7.2s
#14 [stage-0 10/11] RUN printf '\nPATH=/app/node_modules/.bin:$PATH' >> /root/.profile
#14 DONE 0.0s
#15 [stage-0 11/11] COPY . /app
#15 DONE 0.0s
#16 exporting to image
#16 exporting layers
#16 exporting layers 1.2s done
#16 writing image sha256:424e2292378c395b7050c4e3a97c773212a2583408f42afa3bf40c9b14703600 done
#16 naming to docker.io/library/akwk0oscw4woww88gcs04o4w:0de85455a63c48d63aa9aeec367065890f2ced4e-build done
#16 DONE 1.2s
```

**Reply by djsisson** [reply]
yeah i noticed the cache key when looking at the env issue the other day.

one thing that wasn't obvious, is nextjs would do a full rebuild regardless of cache, due to regenerating the code with new id's, however they are introducing a new turbopack persistant cache, so the above methods through nixpacks or manually in a dockerfile would now work great
https://nextjs.org/docs/app/api-reference/config/next-config-js/turbopackPersistentCaching

one thing to add to the future realeases, is an option to specify --cache-from and --cache-to
https://docs.docker.com/reference/cli/docker/buildx/build/#cache-from
This is great for multi build setups, as well as allow the creation of custom builders, especially ones that use a docker container
https://docs.docker.com/reference/cli/docker/buildx/create/#driver

This would be needed for railpack anyway
